frigate/gstreamer.py

```diff
@@ -0,0 +1,359 @@
+            [
+                "gst-discoverer-1.0",
+                "-v",
+                source,
+            ],
+            universal_newlines=True,
+            start_new_session=True,
+            stderr=None,
+            timeout=15,
+        )
+        stripped = list(map(lambda s: s.strip().partition(":"), data.split("\n")))
+        result = {}
+        for key, _, value in stripped:
+            for param in keys:
+                if param == key.lower():
+                    terms = value.strip().split(" ")
+                    result[param] = terms[0].split(",")[0]
+        return result
+    except sp.TimeoutExpired:
+        logger.error(
+            (
+                "gst-discoverer-1.0 timed out auto discovering camera %s. "
+                "Try setting up `decoder_pipeline` according to your camera video codec."
+            ),
+            cam_name,
+        )
+        return {}
+    except:
+        logger.error(
+            "gst-discoverer-1.0 failed with the message: %s", traceback.format_exc()
+        )
+        return {}
+
+
+@lru_cache
+def gst_inspect_find_codec(codec: Optional[str]) -> List[str]:
+    """
+    run gst-inspect-1.0 and find the codec.
+    gst-inspect-1.0 return data in the following format:
+    omx:  omxh265dec: OpenMAX H.265 Video Decoder
+    rtp:  rtph265pay: RTP H265 payloader
+    """
+    try:
+        data = sp.check_output(
+            ["gst-inspect-1.0"],
+            universal_newlines=True,
+            start_new_session=True,
+            stderr=None,
+        )
+        data = [
+            line.split(":")
+            for line in data.split("\n")
+            if codec is None or codec in line
+        ]
+        return [item[1].strip() for item in data if len(item) > 1]
+    except:
+        logger.error(
+            "gst-inspect-1.0 failed with the message: %s", traceback.format_exc()
+        )
+        return None
+
+
+RTP_STREAM_NAME_KEY = "name="
+RTP_STREAM_NAME = "rtp_stream"
+DEPAYED_STREAM_NAME = "depayed_stream"
+
+
+AUDIO_PIPELINES = {
+    "audio/mpeg": ["rtpmp4gdepay", "aacparse"],
+    "audio/x-alaw": ["rtppcmadepay", "alawdec", "audioconvert", "queue", "voaacenc"],
+}
+
+
+class GstreamerBaseBuilder:
+    def __init__(self, width, height, name, format="I420") -> None:
+        self.width = width
+        self.height = height
+        self.name = name
+        self.format = format
+        self.input_pipeline = None
+        self.video_format = None
+        self.record_pipeline = None
+        self.audio_pipeline = None
+        self.raw_pipeline = None
+
+    def with_raw_pipeline(self, raw_pipeline: List[str]):
+        """
+        Set the raw pipeline
+        """
+        self.raw_pipeline = raw_pipeline
+        return self
+
+    def with_source(self, uri: str, options: List[str]):
+        """
+        Set RTMP or RTSP data source with the list of options
+        """
+        is_rtsp = "rtsp://" in uri
+        is_rtmp = "rtmp://" in uri
```

Comment:
UniFi Protect cameras export URLs that are `rtsps://` by default. If it's trivial to support that protocol, it would be nice to add.